examples/split_learning/llm_split_learning/split_learning_trainer.py

```diff
@@ -0,0 +1,168 @@
+    def __init__(
+        self,
+        model,
+        args,
+        train_dataset,
+        eval_dataset,
+        tokenizer,
+        data_collator,
+        sampler,
+        callbacks,
+    ):
+        super().__init__(
+            model=model,
+            args=args,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            tokenizer=tokenizer,
+            data_collator=data_collator,
+            compute_metrics=compute_metrics,
+            preprocess_logits_for_metrics=preprocess_logits_for_metrics,
+            callbacks=callbacks,
+        )
+        self.sampler = sampler
+
+    def _get_train_sampler(self) -> Optional[Sampler]:
+        if self.sampler is None:
+            return RandomSampler(self.train_dataset)
+
+        return self.sampler
+
+    def _get_eval_sampler(self, eval_dataset) -> Optional[Sampler]:
+        if self.sampler is None:
+            return super()._get_eval_sampler(eval_dataset)
+
+        return self.sampler
+
+
+class Trainer(split_learning.Trainer):
+    """The split learning trainer to fine-tune LLM."""
+
+    def __init__(self, model=None, callbacks=None):
+        super().__init__(model, callbacks)
+        # We resize the embeddings to avoid index errors.
+        tokenizer_kwargs = {
+            "cache_dir": Config().params["data_path"],
+            "use_fast": True,
+            "revision": "main",
+            "use_auth_token": None,
+        }
+        model_name = Config().trainer.model_name
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name, **tokenizer_kwargs)
+        embedding_size = self.model.get_input_embeddings().weight.shape[0]
+        if len(self.tokenizer) > embedding_size:
+            self.model.resize_token_embeddings(len(self.tokenizer))
+        # self.training args for huggingface training
+        parser = HfArgumentParser(TrainingArguments)
+        (self.training_args,) = parser.parse_args_into_dataclasses(
+            args=["--output_dir=/tmp", "--report_to=none"]
+        )
+
+    # Redesign the evaluation stage.
+    def test_model_split_learning(self, batch_size, testset, sampler=None):
+        """The testing loop for HuggingFace models.
+
+        Arguments:
+            config: Configuration parameters as a dictionary.
+            testset: The test dataset.
+        """
+        self.training_args.per_device_eval_batch_size = batch_size
+        self.model.copy_weight()
+
+        tester = SampledHuggingFaceTrainer(
+            model=self.model.base_model,
+            args=self.training_args,
+            train_dataset=None,
+            eval_dataset=testset,
+            tokenizer=self.tokenizer,
+            data_collator=default_data_collator,
+            sampler=sampler,
+            callbacks=None,
+        )
+
+        metrics = tester.evaluate()
+
+        # save other metric information such as accuracy
+        tester.log_metrics("eval", metrics)
+        tester.save_metrics("eval", metrics)
```

Comment:
Looks like the eval results will be saved to ''/tmp/eval_results.json''. However, if someone has created this file before and didn't remove it. Other users will encounter a permission denied error since the program tries to overwrite someone else's file.

```
  File "/home/chenghao/miniconda3/envs/plato/lib/python3.9/site-packages/plato/servers/split_learning.py", line 93, in aggregate_weights
    self.test_accuracy = self.trainer.test(self.testset, self.testset_sampler)                                                                                                                
  File "/home/chenghao/miniconda3/envs/plato/lib/python3.9/site-packages/plato/trainers/basic.py", line 396, in test
    accuracy = self.test_process(config, testset, **kwargs)                                                                                                                                   
  File "/home/chenghao/miniconda3/envs/plato/lib/python3.9/site-packages/plato/trainers/basic.py", line 350, in test_process
    raise testing_exception                                                                    
  File "/home/chenghao/miniconda3/envs/plato/lib/python3.9/site-packages/plato/trainers/basic.py", line 344, in test_process
    accuracy = self.test_model(config, testset, **kwargs)                                                                                                                                     
  File "/home/chenghao/miniconda3/envs/plato/lib/python3.9/site-packages/plato/trainers/split_learning.py", line 199, in test_model
    accuracy = self.test_model_split_learning(batch_size, testset, sampler)                                                                                                                   
  File "/home/chenghao/workspace/plato/examples/split_learning/llm_split_learning/split_learning_trainer.py", line 135, in test_model_split_learning
    tester.save_metrics("eval", metrics)                                                       
  File "/home/chenghao/miniconda3/envs/plato/lib/python3.9/site-packages/transformers/trainer_pt_utils.py", line 1017, in save_metrics
    with open(path, "w") as f:                                                                 
PermissionError: [Errno 13] Permission denied: '/tmp/eval_results.json'
```

Reply:
I see, thanks a lot for the review. I think we do not need ```save_metrics``` here very much as I have let the eval_accuracy be reported in the Plato result csv. I simply removed it with a new commit. And I used the ```Config.param[""checkpoint_path"]``` as the temporary path for the Hugging Face trainer.